cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -416,6 +416,109 @@ def _count_matrix_nonzero(self, matrix_name: str, matrix: Union[np.ndarray, spar
+            "genetic_ancestry_East_Asian",
+            "genetic_ancestry_European",
+            "genetic_ancestry_Indigenous_American",
+            "genetic_ancestry_Oceanian",
+            "genetic_ancestry_South_Asian",
+        ]
+
+        organism_column = "organism_ontology_term_id"
+        donor_id_column = "donor_id"
+
+        # Skip any additional validation if the genetic ancestry or organism columns are not present
+        # An error for missing columns will be raised at a different point
+        required_columns = ancestry_columns + [organism_column, donor_id_column]
+        for column in required_columns:
+            if column not in self.adata.obs.columns:
+                return
+
+        donor_id_to_ancestry_values = dict()
+
+        def is_valid_row(row):
+            ancestry_values = row[ancestry_columns]
+
+            # If ancestry values are different for the same donor id, then this row is invalid
+            donor_id = row[donor_id_column]
+            if donor_id in donor_id_to_ancestry_values:
+                if not donor_id_to_ancestry_values[donor_id].equals(ancestry_values):
+                    return False
+            else:
+                donor_id_to_ancestry_values[donor_id] = ancestry_values
+
+            # All values are NaN. This is always valid, regardless of organism
+            if ancestry_values.isna().all():
+                return True
+
+            # If any values are NaN, and we didn't return in the earlier all NaN check, then
+            # this is invalid
+            if ancestry_values.isna().any():
+                return False
+
+            # If organism is not homo sapiens, and we didn't return in the earlier all NaN check,
+            # then this row is invalid
+            if row[organism_column] != "NCBITaxon:9606":
+                return False
+
+            # The sum of genetic ancestry values should be approximately 1.0
+            if (
+                ancestry_values.apply(lambda x: isinstance(x, (float, int))).all()
+                and abs(ancestry_values.sum() - 1.0) <= 1e-6
+            ):
+                return True
+
+            return False
+
+        invalid_rows = ~self.adata.obs.apply(is_valid_row, axis=1)
+
+        if invalid_rows.any():
+            invalid_indices = self.adata.obs.index[invalid_rows].tolist()
+            self.errors.append(
+                f"obs rows with indices {invalid_indices} have invalid genetic_ancestry_* values. All "
+                f"observations with the same donor_id must contain the same genetic_ancestry_* values. If "
+                f"organism_ontolology_term_id is NOT 'NCBITaxon:9606' for Homo sapiens, then all genetic"
+                f"ancestry values MUST be float('nan'). If organism_ontolology_term_id is 'NCBITaxon:9606' "
+                f"for Homo sapiens, then the value MUST be a float('nan') if unavailable; otherwise, the "
+                f"sum of all genetic_ancestry_* fields must be equal to 1.0"
```

Comment:
Yeah I considered that, but it gets a little tricky to implement if you have multiple different violations for each row. I would also expect that in practice it's far more likely that you would see the same type of violation globally (ex: using `None` instead of `float('nan')` for all rows), so having an error logged for each row violation might be excessive. But I'll wait to see what Lattice says in the QA process and if they feel strongly about this.